### ISSUE TITLE
Floating point issues generates wrong totals

### ISSUE THREAD

**Issue by bogdanbradeanu** [bumbummen99/LaravelShoppingcart#142]
Hello,
This problem touches on all versions because this is a PHP floating point behaviour. Usually when dealing with prices all apps must multiply by 100 and deal with Cents not Dollars + Cents.

Example:
Lets say i create a cart with some items, i attach a global discount of 10% and my order total is 60.40 Eur. I save it to Database.
If i recreate the cart in order to refund some items i want to show the difference between the original total and new total. This is how i spotted the problem. The app was showing me that 60.40 Eur - 60.40 !== 0.0
<img width="724" alt="Screen Shot 2021-12-14 at 11 26 32 AM" src="https://user-images.githubusercontent.com/5293776/145970747-479934d2-1cc6-48bf-8233-2ea3b3ced45a.png">
 Eur = 7.105427357601E-15

```
($order->total - ($cart->totalFloat() + $transportValue))
$order->total is of type float 60.40
$cart->totalFloat() is of type float 60.40 (but actually, 60.39 because its rounded up)
$transportValue = 10.00 Eur
```

[https://stackoverflow.com/questions/12291042/weird-php-floating-point-behavior](url)

**Comment by bumbummen99** [comment]
This might be related to #58. AFAIK the issue is that the package does not use absolute values, a solution would be for example to use integers internally or go as far and integrate a lib like MoneyPHP that has a configuration for rounding.

**Comment by stefanocurnis** [comment]
Maybe we could use bcmath.
I use it in my calculator but maybe we can use it also in Cart or CartItem.

Do you want me to try it?

**Comment by bumbummen99** [comment]
> Maybe we could use bcmath. I use it in my calculator but maybe we can use it also in Cart or CartItem.
> 
> Do you want me to try it?

Yep we do the same for our cart package. I already started an refactoring to use [MoneyPHP](https://github.com/moneyphp/money)wich builds on BCMath. That is pretty complete but what is missing from the refactoring is some way to replace the session storage with database or even better abstract it so any storage can be used - this is essential for API projects where there is simply no session.

**Comment by stefanocurnis** [comment]
Oh great! Thanks. Do you plan to release both in a single major version? Or release the MoneyPHP refactor first and then the storage abstraction later?

**Comment by bumbummen99** [comment]
> Oh great! Thanks. Do you plan to release both in a single major version? Or release the MoneyPHP refactor first and then the storage abstraction later?

I was planning to only do the refactor for MoneyPHP but that also does require a workflow change i.e. working with integers and calculation also changed to use MoneyPHP. Eventually i want all calculation, order and so on to be in another package, this should just be a a cart. But at some point i noticed it does not make sense to do a major version without abstracting cart storage, a lot of times people look for a solution that is compatible with API. It would be preferable to be able to choose any storage one would like and to also make it compatible with *mega carts* so that it can be paginated - that is why i wanted to use eloquent for the API/Persistent storage and Session for temporary.
